docs/core/concepts/results.md

```diff
@@ -28,44 +20,116 @@ An instantiated Result object has the following attributes:
+In the following example, imagine a flow was configured to checkpoint each task's output using the `LocalResultHandler` into the `~/Desktop/HelloWorld/results` directory. The `Result.value` is equal to `3` - the actual return value of the Task. The `Result.safe_value` is a string for a local file path.
+
+```python
+>>> state.result[first_result]._result.value
+3
+>>> state.result[first_result]._result.safe_value           
+<SafeResult: '/Users/prefect/Desktop/HelloWorld/results/prefect-result-2020-02-23t18-38-40-381223-00-00'>
+```
+
+Inside that file is a `cloudpickle` of the value `3`, as the implementation of `LocalResultHandler` persists the data from `Result.value` as a pickle at the local filesystem location of `Result.safe_value`.
+```python
+>>> f = open(state.result[first_result]._result.safe_value.value, 'rb')
+>>> content = f.read()     
+>>> content                
+b'\x80\x05K\x03.'
+>>> import cloudpickle     
+>>> cloudpickle.loads(content)                              
+3
+```
+
+However, depending on your situation and the size of your data, you can configure result handlers to serialize a task's entire `Result`, not just its metadata. For example, Core's base `JSONResultHandler` serializes the entire `Result` object as its return value. If you have a task which returns a small piece of data such as a string, or a short list of numbers, handling that data directly in the returned JSON string will suffice!
+
+To showcase this, below is the same example with a flow configured to checkpoint each task's output using the `JSONResultHandler`. Note that now both `Result.value` and `Result.safe_value` contain references to the actual return value of its task, the integer `3`.
+
+```python
+>>> state.result[first_result]._result.value                 
+3
+>>> state.result[first_result]._result.safe_value            
+<SafeResult: '3'>
+```
+
+In Core, the return value is associated in-memory with the `Result` object attached to flow `State`s. For Cloud customers, the return value of a result handler's `write` method is also persisted in Prefect Cloud's database.
 
 ::: warning Handle your data carefully
-When running on Prefect Cloud, the output of a Result Handler's `write` method is what is stored in the Cloud database.
+Using result handlers means that data will be persisted beyond the Prefect's Python process to a storage location, so it is worth taking some extra time to consider what data is safe to persist where.
+
+When running on Prefect Cloud, the return value of a result handler's `write` method is what is stored in the Cloud database. Though you may serialize your data during your result handler's execution, *only* `return` data from Tasks that you feel is safe to persist in Prefect Cloud.
```

Comment:
I think this sounds like the `return` statement within a Task should be checked for safety, when I _think_ you meant that the return value from `write` should be checked.

Reply:
Updated in https://github.com/PrefectHQ/prefect/pull/2076/commits/6ecd02bc9b0a23947b40ad8249258b781a4a1260. Also m it less heavily implied that someone is writing their own result handler.